core/engine/src/job.rs

```diff
@@ -115,6 +116,87 @@ impl NativeJob {
     }
 }
 
+/// An ECMAScript [Job] that runs after a certain amount of time.
+///
+/// This represents the [`HostEnqueueTimeoutJob`] operation from the specification.
+///
+/// [HostEnqueueTimeoutJob]: https://tc39.es/ecma262/#sec-hostenqueuetimeoutjob
+pub struct TimeoutJob {
+    /// The instant this job should be run, in msec since epoch. This will be compared
+    /// to the host's [`HostHooks::utc_now`] method.
+    timeout: i64,
+    /// The job to run after the time has passed.
+    job: NativeJob,
+}
+
+impl Debug for TimeoutJob {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TimeoutJob")
+            .field("timeout", &self.timeout)
+            .field("job", &self.job)
+            .finish()
+    }
+}
+
+impl TimeoutJob {
+    /// Create a new `TimeoutJob` with a timeout and a job.
+    #[must_use]
+    pub fn new_unchecked(job: NativeJob, timeout: i64) -> Self {
+        Self { timeout, job }
+    }
+
+    /// Create a new `TimeoutJob` with a job and a timeout in milliseconds in the future.
+    #[must_use]
+    pub fn delayed(job: NativeJob, delay: u64, context: &mut Context) -> Self {
+        Self::new_unchecked(job, context.host_hooks().utc_now() + (delay as i64))
+    }
+
+    /// Creates a new `TimeoutJob` from a closure and a timeout as [`std::time::SystemTime`].
+    #[must_use]
+    pub fn new<F>(f: F, timeout: std::time::SystemTime) -> Self
+    where
+        F: FnOnce(&mut Context) -> JsResult<JsValue> + 'static,
+    {
+        Self::new_unchecked(
+            NativeJob::new(f),
+            timeout
+                .duration_since(std::time::UNIX_EPOCH)
+                .expect("Invalid SystemTime")
+                .as_millis() as i64,
+        )
+    }
```

Comment:
I think we shouldn't depend so heavily on `SystemTime`. Seems weird to tie one of our job types to something that could panic on certain platforms, and users won't be able to sandbox the engine from the OS if this is in place.

Having `new(job: NativeJob, timeout: i64)` sounds enough for all usecases, and the specific executor can use the provided timeout to schedule the job as it fits.

Reply:
The point of execution needs to be an `Instant`-like point (monotonic and absolute). It cannot be left to interpretation.

Reply:
I would be in favour of using `Duration` instead. The spec is very clear about the milliseconds` argument on [`HostEnqueueTimeoutJob`](https://tc39.es/ecma262/#sec-hostenqueuetimeoutjob):

> It schedules timeoutJob in the realm `realm` in the agent signified by `realm.[[AgentSignifier]]` to be performed after at least `milliseconds` milliseconds.

Hence, it should be fine to use `Duration` in this case.

For `utc_now`, I think we'd have to rethink our approach a bit, maybe doing a survey of what libraries do to support monotonically increasing instants.

Reply:
Done. `TimeoutJob` now only holds the number of millis as a timeout, and it's the JobExecutor's job to add it to the current time and keep that internally.